lib/mainloop-control.c

```diff
@@ -128,45 +144,85 @@ show_ose_license_info(ControlConnection *cc, GString *command, gpointer user_dat
                                 g_string_new("OK You are using the Open Source Edition of syslog-ng."));
 }
 
+static void
+_send_response(gpointer data, gpointer user_data)
+{
+  ControlConnection *cc = (ControlConnection *) data;
+  GString *response = (GString *) user_data;
+  control_connection_send_reply_async(cc, g_string_new(response->str));
+}
+
+static void
+_control_command_async_send_response(ControlCommandAsync *cmd, GString *response)
+{
+  g_mutex_lock(&cmd->args.lock);
+  {
+    g_list_foreach(cmd->args.connections, _send_response, response);
+    g_string_free(response, TRUE);
+    g_list_free(cmd->args.commands);
+    g_list_free(cmd->args.connections);
+    cmd->args.commands = NULL;
+    cmd->args.connections = NULL;
+  }
+  g_mutex_unlock(&cmd->args.lock);
+}
+
 static void
 _respond_config_reload_status(gint type, gpointer user_data)
 {
-  gpointer *args = user_data;
-  MainLoop *main_loop = (MainLoop *) args[0];
-  ControlConnection *cc = (ControlConnection *) args[1];
+  ControlCommandAsync *cmd = (ControlCommandAsync *) user_data;
+  MainLoop *main_loop = cmd->main_loop;
   GString *reply;
 
   if (main_loop_was_last_reload_successful(main_loop))
     reply = g_string_new("OK Config reload successful");
   else
     reply = g_string_new("FAIL Config reload failed, reverted to previous config");
 
-  control_connection_send_reply(cc, reply);
+  _control_command_async_send_response(cmd, reply);
 }
 
 static void
-control_connection_reload(ControlConnection *cc, GString *command, gpointer user_data)
+_control_connection_reload(gpointer user_data)
 {
-  MainLoop *main_loop = (MainLoop *) user_data;
-  static gpointer args[2];
-  GError *error = NULL;
+  main_loop_assert_main_thread();
 
+  ControlCommandAsync *cmd = (ControlCommandAsync *) user_data;
+  MainLoop *main_loop = cmd->main_loop;
+
+  GError *error = NULL;
 
   if (!main_loop_reload_config_prepare(main_loop, &error))
     {
       GString *result = g_string_new("");
       g_string_printf(result, "FAIL %s, previous config remained intact", error->message);
       g_clear_error(&error);
-      control_connection_send_reply(cc, result);
+      _control_command_async_send_response(cmd, result);
       return;
     }
 
-  args[0] = main_loop;
-  args[1] = cc;
-  register_application_hook(AH_CONFIG_CHANGED, _respond_config_reload_status, args);
+  register_application_hook(AH_CONFIG_CHANGED, _respond_config_reload_status, cmd);
   main_loop_reload_config_commence(main_loop);
 }
 
+static void
+control_connection_reload(ControlConnection *cc, GString *command, gpointer user_data)
+{
+  ControlCommandAsync *cmd = (ControlCommandAsync *) user_data;
+  gboolean first_reload = FALSE;
+
+  g_mutex_lock(&cmd->args.lock);
+  {
+    first_reload = cmd->args.commands == NULL;
+    cmd->args.commands = g_list_append(cmd->args.commands, command);
```

Comment:
It's not used currently but after we can have multiple connections, we can have multiple commands.
Otherwise all the changes would be stats-query specific, which would be a bug sooner or later (think of `credentials`, or `config`).